src/App.tsx

```diff
@@ -1,24 +1,219 @@
+  }, []);
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout;
+
+    if (error) {
+      timeout = setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [error]);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const filteredTodos = useMemo(() => {
+    switch (filter) {
+      case StatusFilterType.COMPLETED:
+        return completedTodos;
+
+      case StatusFilterType.ACTIVE:
+        return activeTodos;
+
+      default:
+        return todos;
+    }
+  }, [todos, filter]);
+
+  const handleErrorClose = () => {
+    setError(null);
+  };
+
+  const addTodo = useCallback(async (title: string) => {
+    try {
+      const newTodo = {
+        userId: USER_ID,
+        completed: false,
+        title,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos((prevTodos) => [...prevTodos, createdTodo]);
+    } catch {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+  }, []);
+
+  const removeTodo = useCallback(async (todoId: number) => {
+    try {
+      setLoadingTodoIds(prevTodoId => [...prevTodoId, todoId]);
+      await deleteTodo(todoId);
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete a todo');
+    } finally {
+      setLoadingTodoIds([0]);
+    }
+  }, []);
+
+  const handleClearCompletedTodos = () => {
+    completedTodos.forEach(async (todo) => {
+      await removeTodo(todo.id);
+    });
+  };
+
+  const updateTodo = useCallback(async (
+    todoId: number,
+    args: UpdateTodoArgs,
+  ) => {
+    try {
+      setLoadingTodoIds(prevLoadingTodoIds => [...prevLoadingTodoIds, todoId]);
+      const updatedTodo: Todo = await patchTodo(todoId, args);
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id !== todoId) {
+          return todo;
+        }
+
+        return updatedTodo;
+      }));
+    } catch {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingTodoIds((prevLoadingTodoIds) => (
+        prevLoadingTodoIds.filter((id) => id !== todoId)));
+    }
+  }, []);
+
+  const handleToggleAll = async () => {
+    const allCompleted = todos.every(todo => todo.completed);
+
+    const updatedTodos = todos.map(todo => ({
+      ...todo,
+      completed: !allCompleted,
+    }));
+
+    setTodos(updatedTodos);
+
+    try {
+      setLoadingTodoIds(prevLoadingTodoIds => (
+        [...prevLoadingTodoIds, ...updatedTodos.map(todo => todo.id)]));
+      await Promise.all(updatedTodos.map(todo => (
+        updateTodo(todo.id, { completed: !allCompleted }))));
+    } catch {
+      setError('Unable to update todos');
+    } finally {
+      setLoadingTodoIds([]);
+    }
+  };
```

Comment:
![image](https://github.com/mate-academy/react_todo-app-with-api/assets/84296041/d5eef662-f73e-487e-8fda-9ff3e3e820db)
When I change all todos statuses and I have only 1 uncompleted, I need to send only 1 request, but now it send all todos to update.